### PR TITLE
Disable local notifications with Notification Settings switch

### LINKED ISSUE #15816: Weekly roundup notification despite notifications off
### Expected behavior
When all notifications are off, the weekly roundup notifications are still delivered.

### Actual behavior
The weekly roundup notifications shouldn't be delivered. 

Ideally, there should be a toggle to specifically disable just them. I'd rather have one or two notifications enabled, but toggling 7-8 toggles per site is unnecessarily hard, so I've disabled all notifications at the top of the notifications settings screen. Weekly roundup still came through!

Now I've removed the notification permissions for the app from system settings, but thought I'd let you know just in case this is a bug and not a growth hack.

### Steps to reproduce the behaviour

1. Switch off all notifications using the top toggle on the notification settings screen
2. Wait for weekly roundup notifications

##### Tested on [device], Android [version], WPAndroid [version]
Pixel 5, Android 12, WP 18.8

### PR DESCRIPTION
**Pull request by irfano**
Fixes #15816 

This prevents local notifications when the Notification Settings switch is off. Prevented local notifications are
- Weekly roundup notifications,
- Blogging reminder notifications,
- Create site notifications.

Blogging reminders popups are still active and can be displayed in some cases (eg. from Site Settings) even Notification Settings switch is off. If the user sets a new reminder from the popup, it will be scheduled even if the switch is off.

**Merge requirements:**
- [x] 1. Merge [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2571)
- [x] 2. Update FluxC reference to `trunk`
- [x] 3. Remove `[Status] Not Ready for Merge` label
- [x] 4. Merge this

### To test:
You can wait for scheduled dates to test notifications. It would be the best method but testing all cases takes an unrealistic long time. Alternatively, you can check [App Inspection > Background Task Inspector](https://developer.android.com/studio/inspect/task#inspect-jobs-alarms-wakelocks) on Android Studio if these notifications are scheduled.

**Weekly roundup notifications**
1. Launch the WordPress app. 
2. Navigate to Notifications.
3. Tap ⚙️ button at the top of the screen.
4. Turn off the switch at the top of the page.
5. Verify `WeeklyRoundupWorker` is canceled on `App Inspection > Background Task Inspector`.
6. Turn on the switch again. 
7. Verify `WeeklyRoundupWorker` is scheduled on `App Inspection > Background Task Inspector`.
8. Repeat 1-7 for Jetpack app.

**Blogging reminder notifications**
1. Launch the WordPress app. 
2. Navigate to `My Site → MENU → Site Settings`.
3. Tap Blogging Reminders.
4. Set a reminder.
5. Verify `ReminderWorker` is scheduled on `App Inspection > Background Task Inspector`.
6. Navigate back and change your selected site from the site picker at the top of the MENU page.
7. Navigate to Site Settings and set another reminder for the selected site.
8. Verify a new `ReminderWorker` is scheduled on `App Inspection > Background Task Inspector`.
9. Navigate to Notifications. Tap ⚙️ button at the top of the screen.
10. Turn off the switch at the top of the page.
11. Verify both `ReminderWorker`s are canceled on `App Inspection > Background Task Inspector`.
12. Turn on the switch.
13. Verify both `ReminderWorker`s are scheduled on `App Inspection > Background Task Inspector`.
14. Repeat 1-13 for Jetpack app.

**Create site notifications**
1. Uninstall the app if it's installed.
2. Install and launch the WordPress app. 
3. Log in with an account without any site.
4. Verify `LocalNotificationWorker` is scheduled on `App Inspection > Background Task Inspector`.
5. Navigate to Notifications.
6. Tap ⚙️ button at the top of the screen.
7. Turn off the switch at the top of the page.
8. Verify `LocalNotificationWorker` is canceled on `App Inspection > Background Task Inspector`.
**NOTE:** Create site notifications won't be rescheduled if you turn on the switch again. Two Create site notifications are scheduled when the app is installed the first time. One is for the next day after the install, another one is for 8 days after the first install. If the switch is turned off and turned on in 8 days, create site notifications won't display. I'll discuss this later and change it in a separate PR if necessary.

## Regression Notes
1. Potential unintended areas of impact
Rescheduling blogging reminders after turning Notification Settings off and on again.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually.

3. What automated tests I added (or what prevented me from doing so)
Update `CreateSiteNotificationHandlerTest`.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        setContentView(layout.notifications_settings_activity)
+        fragmentContainer = findViewById(id.fragment_container)
+
+        // Get shared preferences for main switch.
+        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this@NotificationsSettingsActivity)
+
+        // Set up primary toolbar
+        setUpToolbar()
+
+        // Set up main switch
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt]: comment is redundant here

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(layout.notifications_settings_activity)
+        fragmentContainer = findViewById(id.fragment_container)
+
+        // Get shared preferences for main switch.
+        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this@NotificationsSettingsActivity)
+
+        // Set up primary toolbar
+        setUpToolbar()
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt]: Could be renamed to `setupPrimaryToolbar()` and remove the comment

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    private lateinit var sharedPreferences: SharedPreferences
+    private lateinit var fragmentContainer: View
+
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(layout.notifications_settings_activity)
+        fragmentContainer = findViewById(id.fragment_container)
+
+        // Get shared preferences for main switch.
+        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this@NotificationsSettingsActivity)
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt]: Same here, renaming it to mainSwitchPrefs would remove the need for comment

File: RELEASE-NOTES.txt [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 21.3
 -----
+* [*] Disable local notifications with Notification Settings switch. [https://github.com/wordpress-mobile/WordPress-Android/pull/17496]
```
Comment (ravishanker) [inline RELEASE-NOTES.txt:5]: Should this be marked as [internal]

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import org.wordpress.android.modules.APPLICATION_SCOPE
+import org.wordpress.android.ui.LocaleAwareActivity
+import org.wordpress.android.ui.notifications.NotificationEvents.NotificationsSettingsStatusChanged
+import org.wordpress.android.ui.prefs.notifications.PrefMainSwitchToolbarView.MainSwitchToolbarListener
+import org.wordpress.android.ui.prefs.notifications.usecase.UpdateNotificationSettingsUseCase
+import javax.inject.Inject
+import javax.inject.Named
+
+@AndroidEntryPoint
+class NotificationsSettingsActivity : LocaleAwareActivity(), MainSwitchToolbarListener {
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt:33]: 🎉  Kudos for converting it to Kotlin

**Review by ravishanker — APPROVED** [review]
Nice work 👍 

Feel free to merge it after resolving build issues

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        val mainSwitchToolBarView = findViewById<PrefMainSwitchToolbarView>(id.main_switch)
+        mainSwitchToolBarView.setMainSwitchToolbarListener(this)
+
+        // Set main switch state from shared preferences.
+        val isMainChecked = sharedPreferences.getBoolean(getString(string.wp_pref_notifications_main), true)
+        mainSwitchToolBarView.loadInitialState(isMainChecked)
+        hideDisabledView(isMainChecked)
+    }
+
+    override fun onMainSwitchCheckedChanged(buttonView: CompoundButton?, isChecked: Boolean) {
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt:121]: This was removed in a previous PR right?

File: RELEASE-NOTES.txt (hunk 4 above)
Comment (irfano) [reply RELEASE-NOTES.txt:5]: I think an important issue was fixed. This is not only about Jetpack Focus. With this PR, the Notifications Settings switch will affect local notifications too. So, I think it should not be internal.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt (hunk 6 above)
Comment (irfano) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt:121]: No, `onMainSwitchCheckedChanged` was being called unnecessarily at launch. I fixed that in #17498. It didn't change this class.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt (hunk 3 above)
Comment (irfano) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt]: Actually, `sharedPreferences` doesn't have to be a class variable. I moved it to the function. https://github.com/wordpress-mobile/WordPress-Android/pull/17496/commits/7df6d32733173af5f17a60c4880b0e9c5dae6158

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt (hunk 2 above)
Comment (irfano) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt]: Done! https://github.com/wordpress-mobile/WordPress-Android/pull/17496/commits/7df6d32733173af5f17a60c4880b0e9c5dae6158

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt (hunk 1 above)
Comment (irfano) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt]: I just copied Java while converting. But I agree. The code is already explaining itself.
Removed! https://github.com/wordpress-mobile/WordPress-Android/pull/17496/commits/7df6d32733173af5f17a60c4880b0e9c5dae6158

**Pull request merged**